common/network/queues/concurrent_queue.c.inc

```diff
@@ -0,0 +1,126 @@
+    if (self->back == NULL) {                              \
+      lock_handle_unlock(&self->lock);                     \
+      return NULL;                                         \
+    }                                                      \
+    TYPE *data = &self->back->data;                        \
+    lock_handle_unlock(&self->lock);                       \
+    return data;                                           \
+  }
+
+#define DEFINE_FUNCTION_EMPTY_CONCURRENT_QUEUE_OF(TYPE)    \
+  int CONCURRENT_QUEUE_OF_##TYPE##_EMPTY(                  \
+      struct concurrent_queue_of_##TYPE##_s *const self) { \
+    if (self == NULL) return 1;                            \
+    lock_handle_lock(&self->lock);                         \
+    size_t size = self->size;                              \
+    lock_handle_unlock(&self->lock);                       \
+    return size == 0;                                      \
+  }
+
+#define DEFINE_FUNCTION_SIZE_CONCURRENT_QUEUE_OF(TYPE)     \
+  size_t CONCURRENT_QUEUE_OF_##TYPE##_SIZE(                \
+      struct concurrent_queue_of_##TYPE##_s *const self) { \
+    if (self == NULL) return 0;                            \
+    lock_handle_lock(&self->lock);                         \
+    size_t size = self->size;                              \
+    lock_handle_unlock(&self->lock);                       \
+    return size;                                           \
+  }
+
+#define DEFINE_FUNCTION_PUSH_CONCURRENT_QUEUE_OF(TYPE)                      \
+  int CONCURRENT_QUEUE_OF_##TYPE##_PUSH(                                    \
+      struct concurrent_queue_of_##TYPE##_s *const self, TYPE const data) { \
+    if (self == NULL) return 1;                                             \
+    concurrent_queue_node_of_##TYPE##_t *node = malloc(sizeof(*node));      \
+    if (node == NULL) return 1;                                             \
+    node->data = data;                                                      \
+    node->next = NULL;                                                      \
+    lock_handle_lock(&self->lock);                                          \
+    if (self->front == NULL) {                                              \
+      self->front = node;                                                   \
+      self->back = node;                                                    \
+    } else {                                                                \
+      self->back->next = node;                                              \
+      self->back = node;                                                    \
+    }                                                                       \
+    self->size++;                                                           \
+    lock_handle_unlock(&self->lock);                                        \
+    cond_handle_signal(&self->cond);                                        \
```

Comment:
all methods lock; what exact purpose is the condition variable serving?

Reply:
It allows `pop()` to wait for new data without having to continuously poll